### PR TITLE
Exception handling

### PR DESCRIPTION
**Pull request by addev-ashish**
Please let me know if you see any red flags, i will update unit tests and MSAL next

### CODE REVIEW

File: common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+     * {@inheritDoc}
+     * Return the detailed description explaining why the exception is returned back.
+     */
+    @Override
+    public String getMessage() {
+        if (!StringExtensions.isNullOrBlank(super.getMessage())) {
+            return super.getMessage();
+        }
+
+        return "";
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java]: I'm OK with returning `null` here, unless there's a reason `""` is preferable to you

**Comment by iambmelt** [comment]
This looks like a great start - couple of thoughts, observations, questions.

#### Error codes:
What do we expect developers to do with these?

- If we plan to report them only in telemetry or logging, they look fine as-is
- If we expect developers to display them to end users, we should consider `String` formatting that is more easily presented in an app (`i18n`, spaces not `_`, etc)
- If we expect developers to `switch` on them in `catch` blocks, perhaps we use an `enum` or `int` (with a `String` descriptor)<b>**</b>
- To make Intellisense work better, we could consider isolating these code to their own class _or_ giving them a common prefix such that developers can 'browse' them when searching for a specific code:

```java
public static String ERROR_DEVICE_NETWORK_UNAVAILABLE = "Device network unavailable";
public static String ERROR_MALFORMED_URL = "Malformed URL, cannot complete request";
public static String ERROR_MASHED_POTATOES_NO_GRAVY = "These potatoes have no gravy";
```

#### Exceptions & Access to Response Bodies, Headers
How do developers know when to inspect these fields? 

- Because network headers and response bodies are available on the base `class`, it's unclear when you should inspect these values (as _every_ `Exception` potentially contains this payload, although practically they won't) -- would these fields/getters make more sense to expose on the `ServiceException` subclass?
- `SocketTimeouts` are indistinguishable from the default `int` value of `0` -- perhaps we reserve a negative value for this?

<hr>

<b>**</b>Not that developers cannot do this with `String` types, but it becomes a more brittle solution depending on the rigor applied by the consuming dev. For instance:

```java
catch (CommonCoreException e) {
    final msg = e.getMessage();

    switch (msg) {
        case CommonCoreException.ENCRYPTION_ERROR:
            // This is fine, the developer has used a constant.
            // If this *value* of this field changes later this solution continues to work
            break;
        case "decryption_error":
            // Because the developer has taken a dependency on this value
            // this code will break if it is updated later
    }
}
```

File: common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java (hunk 1 above)
Comment (sdabbiru) [reply common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java]: By .NET conventions, APIs should not return null when the return value type is string. I do not know what Java says about this.

**Comment by sdabbiru** [comment]
I am not able to reply to error codes and exceptions & access to response bodies, headers:

Great to see this being discussed, Brian, Ashish.

In general, I do not like wrapping the exceptions with a common exception type. But, since we are already here, error codes (not just the messages) are must have. Also, please include the underlying exception as innerException...

File: common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java (hunk 1 above)
Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java]: Java doesn't share this convention -- the default implementation of `Throwable#getMessage()` returns `null` ([docs](https://developer.android.com/reference/java/lang/Throwable.html#getMessage()))

File: common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    /**
+     * Get the response headers that indicated an error.
+     *
+     * @return The response headers for the network call, null if not initialized.
+     */
+    public HashMap<String, List<String>> getHttpResponseHeaders() {
+        return mHttpResponseHeaders;
+    }
+
+    public void setHttpResponse(HttpWebResponse response) {
```
Comment (heidijinxujia) [inline common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java]: There isn't any constructor taking HttpWebResponse as a parameter in CommonCoreBaseException. Does it mean we need to call setHttpResponse explicitly in other line when creating the CommonCoreBaseException object with http response info?

Comment (addev-ashish) [reply common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java]: Yes, The current PR was just to review the exception layout, I have not plumbed in the function calls yet.

**Comment by addev-ashish** [comment]
@sdabbiru The mapping to code and messages is done in ADAL, are you suggesting that CommonCore throw both errorCode and Message and ADAL picks up error code while MSAL picks up messages?

File: common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java (hunk 1 above)
Comment (addev-ashish) [reply common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java]: fixed

**Comment by sdabbiru** [comment]
@addev-ashish : I am suggesting that the contract (between layers) should never be such that the client interprets or compares strings.

File: common/src/main/java/com/microsoft/identity/common/exception/CommonCoreServiceException.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.exception;
+
+public class CommonCoreServiceException extends CommonCoreBaseException {
+    /**
+     * This request is missing a required parameter, includes an invalid parameter, includes a parameter more than
+     * once, or is otherwise malformed.
+     */
+    public static final String INVALID_REQUEST = "invalid_request";
```
Comment (heidijinxujia) [inline common/src/main/java/com/microsoft/identity/common/exception/CommonCoreServiceException.java]: Would it be better to group the error code String into an enum in the advantages of type safety, switch-case statements, serialization etc.? The name of properties already explain themselves.

**Comment by addev-ashish** [comment]
@sdabbiru i would agree but that would mean an explosion of exception classes.

File: common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        if (!StringExtensions.isNullOrBlank(super.getMessage())) {
+            return super.getMessage();
+        }
+
+        return null;
+    }
+
+    protected int mHttpStatusCode;
+
+    protected HashMap<String, String> mHttpResponseBody = null;
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java]: This field, `mHttpStatusCode` and `mHttpResponseHeaders` seem like they would be more suited on the `CommonCoreServiceException` subclass -- why does the base class contain these fields? It doesn't seem right to me that a client-side exception would have these fields

File: common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+import com.microsoft.identity.common.adal.internal.net.HttpWebResponse;
+import com.microsoft.identity.common.adal.internal.util.StringExtensions;
+import com.microsoft.identity.common.adal.internal.util.HashMapExtensions;
+
+import org.json.JSONException;
+
+import java.util.HashMap;
+import java.util.List;
+
+public class CommonCoreBaseException extends Exception {
```
Comment (shoatman) [inline common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java]: I suggest removing the CommonCore prefix.  The package name tells you the origin of the exception.

File: common/src/main/java/com/microsoft/identity/common/exception/CommonCoreExceptionMessage.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.exception;
+
+public class CommonCoreExceptionMessage {
```
Comment (shoatman) [inline common/src/main/java/com/microsoft/identity/common/exception/CommonCoreExceptionMessage.java]: These are strings describing error conditions and not specific client and/or server exceptions.  Is that correct?  If so, i'd suggest changing the name to reflect that.

File: common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java (hunk 4 above)
Comment (shoatman) [reply common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java]: +1

File: common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryOAuth2Strategy.java [hunk 7]
```diff
@@ -90,26 +93,20 @@ public String getIssuerCacheIdentifier(AuthorizationRequest request) {
...
     @Override
     public AccessToken getAccessTokenFromResponse(TokenResponse response) {
-        AzureActiveDirectoryAccessToken accessToken;
-
-        if (response instanceof AzureActiveDirectoryTokenResponse) {
-            accessToken = new AzureActiveDirectoryAccessToken(response);
-        } else {
-            throw new RuntimeException("Expected AzureActiveDirectoryTokenResponse in AzureActiveDirectoryOAuth2Strategy.getAccessTokenFromResponse");
+        if (!(response instanceof AzureActiveDirectoryTokenResponse)) {
+            throw new IllegalArgumentException(
```
Comment (shoatman) [inline common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryOAuth2Strategy.java]: Yep.  This is better.

File: common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/ClientInfo.java [hunk 8]
```diff
@@ -40,7 +42,7 @@ public ClientInfo(String rawClientInfo) {
         try {
             clientInfoItems = JsonExtensions.extractJsonObjectIntoMap(decodedClientInfo);
         } catch (final JSONException e) {
-            throw new RuntimeException(e);
+            throw new CommonCoreServiceException("", CommonCoreExceptionMessage.INVALID_JWT, e);
```
Comment (shoatman) [inline common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/ClientInfo.java]: Here's an example that i'd like us to review.  Originally this was a client exception in MSAL.  Is it our expectation that malformed JSON returned from the server would be correct on a subsequent request?  How does the developer know that an invalid JWT is something that they can recover from?  Do they need to make a new silent or a new interactive request?

File: common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryOAuth2Strategy.java [hunk 9]
```diff
@@ -49,23 +49,26 @@ protected void validateTokenRequest(TokenRequest request) {
      */
     @Override
     public Account createAccount(TokenResponse response) {
-        IDToken idToken = new IDToken(response.getIdToken());
-        ClientInfo clientInfo = new ClientInfo(((AzureActiveDirectoryTokenResponse) response).getClientInfo());
+        IDToken idToken = null;
+        ClientInfo clientInfo = null;
+        try {
+            idToken = new IDToken(response.getIdToken());
```
Comment (shoatman) [inline common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryOAuth2Strategy.java]: I'm wondering whether it makes more sense to  validate the entire response prior to attempting to use it in various classes.   In that way... we avoid failing at some point during the processing of the response.  And we avoid the need to handle exceptions all over the place related to parsing, etc....  We validate/parse it in a single step such that it's prepared for further use.

File: common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java (hunk 4 above)
Comment (addev-ashish) [reply common/src/main/java/com/microsoft/identity/common/exception/CommonCoreBaseException.java]: you guys are right, it was just an attempt at avoiding exposing the 2 child classes to ADAL/MSAL. What do you guys think about the tradeoff?

File: common/src/main/java/com/microsoft/identity/common/exception/ServiceException.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+import com.microsoft.identity.common.adal.internal.util.HashMapExtensions;
+
+import org.json.JSONException;
+
+import java.util.HashMap;
+import java.util.List;
+
+public class ServiceException extends BaseException {
+
+    protected int mHttpStatusCode;
```
Comment (heidijinxujia) [inline common/src/main/java/com/microsoft/identity/common/exception/ServiceException.java]: Why `protected`, not `private`? There are getters, right?

File: common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/IDToken.java [hunk 11]
```diff
@@ -180,15 +182,15 @@ public IDToken(final String rawIdToken) {
...
+    private Map<String, String> parseJWT(final String idToken) throws ServiceException {
         final String idTokenBody = extractJWTBody(idToken);
         final byte[] data = Base64.decode(idTokenBody, Base64.URL_SAFE);
 
         try {
             final String decodedBody = new String(data, Charset.forName(StringExtensions.ENCODING_UTF8));
             return JsonExtensions.extractJsonObjectIntoMap(decodedBody);
         } catch (final JSONException e) {
-            throw new RuntimeException(e);
+            throw new ServiceException("", ErrorStrings.INVALID_JWT, e);
```
Comment (heidijinxujia) [inline common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/IDToken.java]: Could error code string be empty?

File: common/src/main/java/com/microsoft/identity/common/exception/ErrorStrings.java [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+// THE SOFTWARE.
+package com.microsoft.identity.common.exception;
+
+public class ErrorStrings {
+    /**
+     * There are multiple cache entries found, the sdk cannot pick the correct access token
+     * or refresh token from the cache. Likely it's a bug in the sdk when caching tokens or authority
+     * is not proviced in the silent request and multiple tokens were found.
+     */
+    public static final String MULTIPLE_MATCHING_TOKENS_DETECTED = "multiple_matching_tokens_detected";
```
Comment (heidijinxujia) [inline common/src/main/java/com/microsoft/identity/common/exception/ErrorStrings.java]: One question, as the ErrorString variables are all String type. What if the dev passed in invalid error String, which is not defined in the ErrorStrings class? Why enum is not preferred?

File: common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/IDToken.java (hunk 11 above)
Comment (addev-ashish) [reply common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/IDToken.java]: keeping it empty for now, we need a diagnosibility push in the end.

File: common/src/main/java/com/microsoft/identity/common/exception/ErrorStrings.java (hunk 12 above)
Comment (addev-ashish) [reply common/src/main/java/com/microsoft/identity/common/exception/ErrorStrings.java]: enums leave our partners comparing large formatted strings, i wanted strings to simplify their life. There will be more error strings exposed as we push for showing more internal errors. I do not have a strong preference per se though.

File: common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/IDToken.java (hunk 11 above)
Comment (shoatman) [reply common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/IDToken.java]: I'm still wondering if this is recoverable.... what are we communicating to the developer here?  What are they supposed to do to resolve this?

**Pull request merged**